src/main/scala/org/broadinstitute/dsde/firecloud/utils/EnabledUserDirectives.scala

```diff
@@ -0,0 +1,126 @@
+          throwErrorReport(code, message)
+        case Failure(ex) =>
+          logger.error(s"Unexpected exception checking enabled status for user ${userInfo.userEmail}: (${ex.getMessage}) while calling $uri", ex)
+          throwErrorReport(StatusCodes.InternalServerError, ex.getMessage)
+      }
+    }
+  }
+
+
+  private class SamApiCallback[T](functionName: String = "userStatusInfo") extends ApiCallback[T] {
+    private val promise = Promise[T]()
+
+    override def onFailure(e: ApiException, statusCode: Int, responseHeaders: java.util.Map[String, java.util.List[String]]): Unit = {
+      val response = e.getResponseBody
+      // attempt to propagate an ErrorReport from Sam. If we can't understand Sam's response as an ErrorReport,
+      // create our own error message.
+      import org.broadinstitute.dsde.rawls.model.WorkspaceJsonSupport.ErrorReportFormat
+      toFutureTry(Unmarshal(response).to[ErrorReport]) flatMap {
+        case Success(err) =>
+          logger.error(s"Sam call to $functionName failed with error $err")
+          throw new FireCloudExceptionWithErrorReport(err)
+        case Failure(_) =>
+          // attempt to extract something useful from the response entity, even though it's not an ErrorReport
+          toFutureTry(Unmarshal(response).to[String]) map { maybeString =>
+            val stringErrMsg = maybeString match {
+              case Success(stringErr) => stringErr
+              case Failure(_) => response
+            }
+            throw new FireCloudExceptionWithErrorReport(ErrorReport(StatusCode.int2StatusCode(statusCode), s"Sam call to $functionName failed with error '$stringErrMsg'"))
+          }
+      }
+      promise.failure(e)
+    }
+    override def onSuccess(result: T, statusCode: Int, responseHeaders: java.util.Map[String, java.util.List[String]]): Unit = promise.success(result)
+
+    override def onUploadProgress(bytesWritten: Long, contentLength: Long, done: Boolean): Unit = ()
+
+    override def onDownloadProgress(bytesRead: Long, contentLength: Long, done: Boolean): Unit = ()
+
+    def future: Future[T] = promise.future
+  }
+
+  private def getUserEnabled(token: String, samBaseUrl: String): Future[Boolean] = {
+    val apiClient = new ApiClient()
+    apiClient.setAccessToken(token)
+    apiClient.setBasePath(samBaseUrl)
+    val sam = new UsersApi(apiClient)
+    val callback = new SamApiCallback[UserStatusInfo]
+    // TODO: should we use retries here?
+    /* TODO: instead of directly calling Sam's userStatusInfo API, should we call a "real" api and, if that api returns
+        401, echo its response? That would allow Orch to always give the same error response that Sam gives, but
+        it is likely to be more expensive in the common case of the user being enabled.
+     */
+    sam.getUserStatusInfoAsync(callback)
+    // TODO: is it enough to check this one `getEnabled` boolean?
+    callback.future map (userStatus => Boolean.unbox(userStatus.getEnabled))
```

Comment:
FWIW that's what I independently decided I should do in CromIAM, not 1000% sure it's right